File: lib/checker.js
'use strict';

const fs = require('fs');
const path = require('path');


/**
 * Checks that the file extension was PNG.
 * @param {string} imgPath
 * @returns {Array|{index: number, input: string}|*}
 */

// @todo Create better checking for PNG, for example exif or a unique byte.
exports.isPNG = function (imgPath) {
    return imgPath.match(/.png$/i);
};


/**
 * Check the file path to the URI.
 * @param {string} imgPath
 * @returns {Array|{index: number, input: string}|*}
 */
exports.isURI = function (imgPath) {
    const url = require('url');
    return url.parse(imgPath).protocol;
};


/**
 * Checks read access to the file.
 * @param {string} imgPath
 * @param callback
 */
exports.hasAccess = function (imgPath, callback) {
    imgPath = path.join(process.cwd(), imgPath);
    fs.access(imgPath, fs.constants.R_OK, (err) => {
        callback(err);
    });
};


/**
 * Checks the syntax of the pattern. Returns an array of constant step length and height.
 * @param step
 * @returns {Array}
 */
exports.isRightStep = function (step) {
    const err = new SyntaxError('Wrong step input. Check help!');
    const divider = step.match(/x|х/gi)[0];
    const pattern = step.split(divider);
    if (pattern.length == 2) {
        return pattern.map(item => {
            if (isNaN(item)) throw err;
            return item
        })
    } else {
        throw err;
    }
};

File: src/checker.rs
use regex::Regex;
use std::error::Error;
use std::fs;
use std::path::Path;
use url::Url;

/// Checks that the file extension is PNG.
pub fn is_png(img_path: &str) -> bool {
  let re = Regex::new(r"(?i)\.png$").unwrap();
  re.is_match(img_path)
}

/// Checks if the given path is a valid HTTP or HTTPS URI.
pub fn is_uri(img_path: &str) -> bool {
  match Url::parse(img_path) {
    Ok(url) => url.scheme() == "http" || url.scheme() == "https",
    Err(_) => false,
  }
}

/// Checks read access to the file.
pub async fn has_access(img_path: &str) -> Result<String, Box<dyn Error>> {
  let path = Path::new(&img_path).canonicalize()?;
  fs::metadata(&path)?;
  Ok(path.to_string_lossy().into_owned())
}

/// Checks the syntax of the pattern. Returns a tuple of step width and height.
pub fn is_right_step(step: &str) -> Result<(String, String), &'static str> {
  let parts: Vec<&str> = step.split(['x', 'х'].as_ref()).collect();
  if parts.len() == 2 {
    let width = parts[0].parse::<i32>().map_err(|_| "Invalid width")?;
    let height = parts[1].parse::<i32>().map_err(|_| "Invalid height")?;
    Ok((width.to_string(), height.to_string()))
  } else {
    Err("Wrong step input. Check help!")
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn test_is_png() {
    assert!(is_png("image.png"));
    assert!(!is_png("image.jpg"));
  }

  #[test]
  fn test_is_uri() {
    assert!(is_uri("http://example.com"));
    assert!(is_uri("https://example.com"));
    assert!(!is_uri("ftp://example.com"));
    assert!(!is_uri("not a uri"));
  }

  #[tokio::test]
  async fn test_has_access() {
    let result = has_access("Cargo.toml").await;
    assert!(result.is_ok());

    let result = has_access("nonexistent_file.png").await;
    assert!(result.is_err());
  }

  #[test]
  fn test_is_right_step() {
    assert_eq!(
      is_right_step("10x20").unwrap(),
      ("10".to_string(), "20".to_string())
    );
    assert_eq!(
      is_right_step("10х20").unwrap(),
      ("10".to_string(), "20".to_string())
    );
    assert!(is_right_step("10x").is_err());
    assert!(is_right_step("x20").is_err());
  }
}
